File: app_pasos/lib/models/challenge.dart
import 'user.dart';

class Challenge {
  final String id;
  final String code;
  final String status;
  final DateTime startDate;
  final User? creator;
  final User? opponent;

  Challenge({
    required this.id,
    required this.code,
    required this.status,
    required this.startDate,
    this.creator,
    this.opponent,
  });

  factory Challenge.fromJson(Map<String, dynamic> json) {
    return Challenge(
      id: json['id'] ?? json['_id'] ?? '',
      code: json['code'] ?? '',
      status: json['status'] ?? 'waiting',
      startDate: json['startDate'] != null
          ? DateTime.parse(json['startDate'])
          : DateTime.now(),
      creator: json['creator'] != null ? User.fromJson(json['creator']) : null,
      opponent: json['opponent'] != null ? User.fromJson(json['opponent']) : null,
    );
  }
}

File: backend/models/Challenge.js
const mongoose = require('mongoose');

const challengeSchema = new mongoose.Schema({
  code: {
    type: String,
    required: true,
    unique: true,
    uppercase: true
  },
  creator: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  opponent: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    default: null
  },
  duration: {
    type: Number,
    default: 30,
    min: 1,
    max: 365
  },
  status: {
    type: String,
    enum: ['waiting', 'active', 'finished'],
    default: 'waiting'
  },
  startDate: {
    type: Date,
    default: Date.now
  },
  endDate: {
    type: Date
  }
}, {
  timestamps: true
});

challengeSchema.index({ creator: 1, status: 1 });
challengeSchema.index({ opponent: 1, status: 1 });

module.exports = mongoose.model('Challenge', challengeSchema);
